### DISCUSSION TITLE
How to embed into a (non Spring based) java web application

### DISCUSSION THREAD

**Discussion by surajamin** [sevdokimov/log-viewer#174] [Q&A]
Hi,

I would like to embed LogViewer into my existing java web application (which is not based on Spring framework). 

Can someone guide me on how this can be achieved? This could then be fed into https://github.com/sevdokimov/log-viewer/blob/master/_docs/embadded.md#to-do-write-documentation

I had a look at the source code of com.logviewer.web.LogViewerServlet and it seems like it needs a Spring context.

Many thanks,

Suraj

**Comment by sevdokimov** [comment]
Hello,

There is an example in https://github.com/sevdokimov/log-viewer/tree/master/demo/simple-java-web-application
The application contains a [servlet](https://github.com/sevdokimov/log-viewer/blob/master/demo/simple-java-web-application/src/main/java/com/logviewer/demo/MyLogViewerServlet.java) that extends `com.logviewer.web.LogViewerServlet`. The init method of that servlet creates a spring context and sets it to `com.logviewer.data2.LogContextHolder#instance`. `LogViewerServlet` uses  `LogContextHolder.getInstance()` to get current spring context.

**Reply by surajamin** [reply]
Hi,

Thanks for your response. I’ve managed to create a very simple war file to be deployed onto Tomcat using your MyLogViewerServlet as a starting point.

Can you please let me know how I can configure with various log-viewer settings? I have a config.conf configured and it works well when I use it with your standalone (jetty based) webapp. 

Do I need to place the configured config.conf somewhere in the web server/classpath or pass it as a system property?

Many thanks,

Suraj

**Reply by sevdokimov** [reply]
It should automatically detect the logs files where your application write logs, a configuration is not needed. If you need to add additional log files, you can create a spring bean implementing `LogConfigurationLoader` interface. Example: 
```
    @Override
    public void init() {
        AnnotationConfigApplicationContext ctx = new AnnotationConfigApplicationContext(LogViewerAutoConfig.class, LvConfigBase.class, MyConfig.class);
        LogContextHolder.setInstance(ctx);

        super.init();
    }

    @Configuration
    public static class MyConfig {
        @Bean
        public LogConfigurationLoader customLogDetector() {
            return new CustomLogDetector();
        }
    }

    public static class CustomLogDetector implements LogConfigurationLoader {
        @Override
        public Map<Path, LogFormat> getLogConfigurations() {
            return Collections.singletonMap(
                    Paths.get("/tmp/additional-log.log"),
                    new Log4jLogFormat("%d{yyyyMMddHHmmss.SSS} %m%n")
            );
        }
    }
```

**Reply by surajamin** [reply]
Hi,

Thanks for your prompt response. 

Is it possible to reuse what you have done already (`TypesafePropertySourceFactory`) so the configuration can be picked up from a config file specified by `log-viewer.config-file` system property?


Let me give a bit of additional context. Your standalone (jetty based) web app would have been ideal to use as-is (with the configurations specified in your config.conf) but we already have Tomcat running on the server and we can't run any other servers (or have any non-standard ports accessible via the firewall). We have several web apps (all non-spring based) writing to different (Log4j based) logs in different folders. So it would be great to have one web app dedicated to providing web-based access to all the logs.

So what I am essentially trying to do is create a LogViewer (tomcat) war version of your standalone app which can look at logs (based on HOCON config) created by other apps/system processes.
  
Having a Logviewer.war that can be deployed to any web app server would be a great addition to this project.


Many thanks,

Suraj
